src/babyjubjub-ecdsa/proofs.ts

```diff
@@ -0,0 +1,296 @@
+    }
+
+    const merkleRoot = getMerkleRootFromCache(this.collectionPubKeys);
+
+    return await verifyMembership({
+      proof: membershipProof,
+      merkleRoot,
+      sigNullifierRandomness: this.sigNullifierRandomness,
+      usedSigNullifiers,
+      pathToCircuits: this.pathToCircuits,
+    });
+  }
+}
+
+export class NUniqueJubmojisInCollection {
+  collectionPubKeys: string[];
+  sigNullifierRandomness: bigint;
+  N: number;
+  pathToCircuits?: string;
+
+  constructor(
+    collectionPubKeys: string[],
+    sigNullifierRandomness: bigint,
+    N: number,
+    pathToCircuits?: string
+  ) {
+    this.collectionPubKeys = collectionPubKeys;
+    this.sigNullifierRandomness = sigNullifierRandomness;
+    this.N = N;
+    this.pathToCircuits = pathToCircuits;
+  }
+
+  // Todo: Let's use batch proving here
+  async prove(jubmojis: JubmojiData[]): Promise<MembershipProof[]> {
+    const pubKeyNullifierRandomness = getRandomNullifierRandomness();
+
+    return await Promise.all(
+      jubmojis.map(
+        async ({
+          jubmojiIndex,
+          sig,
+          msgNonce,
+          msgRand,
+          R,
+          T,
+          U,
+        }: JubmojiData) => {
+          const decodedSig = derDecode(sig);
+          const msgHash = hashMessage(
+            recoverMessageFromNonceAndRand(msgNonce, msgRand)
+          );
+          const pubKey = getJubmojiPubKeyFromFullCollectionIndex(jubmojiIndex);
+          const index = this.collectionPubKeys.indexOf(pubKey);
+          const merkleProof = getMerkleProofFromCache(
+            this.collectionPubKeys,
+            index
+          );
+
+          return await proveMembership({
+            sig: decodedSig,
+            msgHash,
+            publicInputs: {
+              R,
+              T,
+              U,
+            },
+            merkleProof,
+            sigNullifierRandomness: this.sigNullifierRandomness,
+            pubKeyNullifierRandomness,
+            pathToCircuits: this.pathToCircuits,
+          });
+        }
+      )
+    );
+  }
+
+  // Checks that all the membershipProofs are valid, and that their pubKeyNullifiers are unique
+  // while having the same pubKeyNullifierRandomnessHash
+  async verify({
+    membershipProofs,
+    usedSigNullifiers,
+  }: {
+    membershipProofs: MembershipProof[];
+    usedSigNullifiers: bigint[];
+  }): Promise<{
+    verified: boolean;
+    newSigNullifiers?: bigint[];
+  }> {
+    // Must have at least N proofs
+    if (membershipProofs.length < this.N) {
+      return { verified: false };
+    }
+
+    // Checks that all the pubKeyNullifiers are unique but all the
+    // pubKeyNullifierRandomnessHashes are the same
+    const pubKeyNullifiers = membershipProofs.map(
+      (proof) => getPublicSignalsFromMembershipZKP(proof.zkp).pubKeyNullifier
+    );
+    if (!areAllBigIntsDifferent(pubKeyNullifiers)) {
+      return { verified: false };
+    }
+
+    const pubKeyNullifierRandomnessHashes = membershipProofs.map(
```

Comment:
Ah, I see that you just parse the pubKeyNullifier related business here instead of at the verify step. Okay you can just ignore that comment in the BBJJ repo